crates/flux-fhir-analysis/src/wf/sortck.rs

```diff
@@ -84,6 +84,39 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
         }
     }
 
+    fn check_constructor(
+        &mut self,
+        path: &fhir::PathExpr,
+        field_exprs: &[fhir::FieldExpr],
+        _spread: &Option<fhir::Spread>,
+        expected: &rty::Sort,
+    ) -> Result {
+        if let rty::Sort::App(rty::SortCtor::Adt(sort_def), sort_args) = expected {
+            // these are ordered
+            let field_names = sort_def.field_names();
+            let sorts = sort_def.field_sorts(sort_args);
+            for expr in field_exprs {
+                // make sure that the field is actually a field
+                if let Some(idx) = field_names.iter().position(|s| *s == expr.ident.name) {
+                    // check that field sort is the same as the expr sort
+                    let sort = &sorts[idx];
+                    self.check_expr(&expr.expr, sort)?
+                } else {
+                    // emit an error because we have a mismatched field
+                    return Err(self.emit_err(errors::InvalidFieldUpdate::new(expr.ident.span)));
+                }
+            }
+            Ok(())
+        } else {
+            Err(self.emit_err(errors::ArgCountMismatch::new(
+                Some(path.span),
+                String::from("type"),
+                1,
+                field_exprs.len(),
+            )))
```

Comment:
We'll hit this with an expression like `if (S { x: 0}) { 0 } else { 1 }` or `S { x: 0} > 0`. We should report an error saying something like `expected bool, found sort constructor`. This is similar to the `UnexpectedFun` error.